src/lib/operators/join_index.cpp

```diff
@@ -0,0 +1,346 @@
+  _pos_list_right = std::make_shared<PosList>();
+
+  for (ChunkID chunk_id_left = ChunkID{0}; chunk_id_left < _left_in_table->chunk_count(); ++chunk_id_left) {
+    // initialize the data structures for left matches
+    _left_matches[chunk_id_left].resize(_left_in_table->get_chunk(chunk_id_left)->size());
+  }
+
+  // Scan all chunks for right input
+  for (ChunkID chunk_id_right = ChunkID{0}; chunk_id_right < _right_in_table->chunk_count(); ++chunk_id_right) {
+    const auto chunk_right = _right_in_table->get_chunk(chunk_id_right);
+    auto column_right = chunk_right->get_column(_right_column_id);
+    const auto indices = chunk_right->get_indices(std::vector<ColumnID>{_right_column_id});
+    _right_matches[chunk_id_right].resize(chunk_right->size());
+
+    std::shared_ptr<BaseIndex> index = nullptr;
+
+    if (indices.size() > 0) {
+      // We assume the first index to be efficient for our join
+      // as we do not want to spend time on evaluating the best index inside of this join loop
+      index = indices.front();
+    }
+
+    // Scan all chunks from left input
+    for (ChunkID chunk_id_left = ChunkID{0}; chunk_id_left < _left_in_table->chunk_count(); ++chunk_id_left) {
+      auto chunk_column_left = _left_in_table->get_chunk(chunk_id_left)->get_column(_left_column_id);
+
+      // for Outer joins, remember matches on the left side
+      auto& left_matches = _left_matches[chunk_id_left];
+
+      if (_mode == JoinMode::Left || _mode == JoinMode::Outer) {
+        left_matches.resize(chunk_column_left->size());
+      }
+
+      resolve_data_and_column_type(left_data_type, *chunk_column_left, [&](auto left_type, auto& typed_left_column) {
+        resolve_data_and_column_type(right_data_type, *column_right, [&](auto right_type, auto& typed_right_column) {
+          using LeftType = typename decltype(left_type)::type;
+          using RightType = typename decltype(right_type)::type;
+
+          // make sure that we do not compile invalid versions of these lambdas
+          constexpr auto left_is_string_column = (std::is_same<LeftType, std::string>{});
+          constexpr auto right_is_string_column = (std::is_same<RightType, std::string>{});
+
+          constexpr auto neither_is_string_column = !left_is_string_column && !right_is_string_column;
+          constexpr auto both_are_string_columns = left_is_string_column && right_is_string_column;
+
+          // clang-format off
+          if constexpr (neither_is_string_column || both_are_string_columns) {
+              auto iterable_left = create_iterable_from_column<LeftType>(typed_left_column);
+              auto iterable_right = create_iterable_from_column<RightType>(typed_right_column);
+
+              iterable_left.with_iterators([&](auto left_it, auto left_end) {
+                  iterable_right.with_iterators([&](auto right_it, auto right_end) {
+                      with_comparator(_predicate_condition, [&](auto comparator) {
```

Comment:
Same as above. Even if you keep the fallback, the current solution compiles much more code than is necessary. The lambda passed to `with_operator` method is compiled for each operator type, yet the largest part of `_join_two_columns`, where the index is used, doesn’t need the comparator at all. The same is true for the right iterators. So if you decide to keep the fallback, you should still move the index part out of `_join_two_columns` and into the scope of `iterable_left.with_iterators()` so that the same code isn’t compiled multiple times.

Reply:
I agree that the if based on whether `index` is null should be pulled out.